librad/src/net/protocol/interrogation/xor.rs

```diff
@@ -0,0 +1,233 @@
+
+    pub fn try_from_iter<T, E>(iter: T) -> Result<Self, BuildError<E>>
+    where
+        T: IntoIterator<Item = Result<SomeUrn, E>>,
+        E: std::error::Error + Send + Sync + 'static,
+    {
+        let mut xs = Vec::with_capacity(MaxElements::USIZE);
+        for x in iter {
+            let urn = x?;
+            xs.push(xor_hash(&urn));
+            if xs.len() > MaxElements::USIZE {
+                return Err(BuildError::TooManyElements);
+            }
+        }
+
+        let inner = Xor16::from(xs);
+        Ok(Self { inner })
+    }
+}
+
+impl Clone for Xor {
+    fn clone(&self) -> Self {
+        Self {
+            inner: Xor16 {
+                seed: self.inner.seed,
+                block_length: self.inner.block_length,
+                fingerprints: self.inner.fingerprints.clone(),
+            },
+        }
+    }
+}
+
+impl<N> From<&SVec<N, SomeUrn>> for Xor
+where
+    N: Unsigned + IsLessOrEqual<MaxElements>,
+{
+    fn from(svec: &SVec<N, SomeUrn>) -> Self {
+        let inner = Xor16::from(svec.iter().map(xor_hash).collect::<Vec<_>>());
+        Self { inner }
+    }
+}
+
+impl minicbor::Encode for Xor {
+    fn encode<W: minicbor::encode::Write>(
+        &self,
+        e: &mut minicbor::Encoder<W>,
+    ) -> Result<(), minicbor::encode::Error<W::Error>> {
+        Encode {
+            seed: self.inner.seed,
+            block_length: self.inner.block_length,
+            fingerprints: &self.inner.fingerprints,
+        }
+        .encode(e)
+    }
+}
+
+impl<'b> minicbor::Decode<'b> for Xor {
+    fn decode(d: &mut minicbor::Decoder) -> Result<Self, minicbor::decode::Error> {
+        let Decode {
+            seed,
+            block_length,
+            fingerprints,
+        } = minicbor::Decode::decode(d)?;
+        Ok(Self {
+            inner: Xor16 {
+                seed,
+                block_length,
+                fingerprints: fingerprints.into_boxed_slice(),
+            },
+        })
+    }
+}
+
+#[derive(minicbor::Encode)]
+struct Encode<'a> {
+    #[n(0)]
+    seed: u64,
+    #[n(1)]
+    block_length: usize,
+    #[n(2)]
+    fingerprints: &'a [u16],
+}
+
+#[derive(minicbor::Decode)]
+struct Decode {
+    #[n(0)]
+    seed: u64,
+    #[n(1)]
+    block_length: usize,
+    #[n(2)]
+    #[cbor(with = "bounded")]
+    fingerprints: Vec<u16>,
+}
+
+mod bounded {
+    use super::MAX_FINGERPRINTS;
+
+    pub fn decode(d: &mut minicbor::Decoder) -> Result<Vec<u16>, minicbor::decode::Error> {
+        use minicbor::decode::{Decode, Error::Message as Error};
+
+        match d.probe().array()? {
+            None => Err(Error("expected definite-length array")),
+            Some(len) => {
+                if len > MAX_FINGERPRINTS {
+                    Err(Error("max length exceeded"))
+                } else {
+                    Ok(Decode::decode(d)?)
+                }
+            },
+        }
+    }
+}
+
+fn xor_hash(urn: &SomeUrn) -> u64 {
+    let SomeUrn::Git(Urn { id, path: _ }) = urn;
```

Comment:
I'd prefer we got a compiler error from a missing case in `match` if we added a new variant :)

Reply:
But you will - Rust won’t accept refutable patterns in a let binding.

Reply:
Right, I mixed this up with `if let` :facepalm:

Reply:
No no, admit it: you were playing with this Huskall thing again!